tokio/src/runtime/task/harness.rs

```diff
@@ -227,7 +210,58 @@ where
     }
 
     fn to_task(&self) -> Task<S> {
-        unsafe { Task::from_raw(self.header().into()) }
+        self.scheduler_view().to_task()
+    }
+}
+
+struct SchedulerView<'a, S> {
+    header: &'a Header,
+    scheduler: &'a Scheduler<S>,
+}
+
+impl<'a, S> SchedulerView<'a, S>
+where
+    S: Schedule,
+{
+    fn to_task(&self) -> Task<S> {
+        // SAFETY The header is from the same struct containing the scheduler `S` so  the cast is safe
+        unsafe { Task::from_raw(self.header.into()) }
+    }
+
+    fn transition_to_running(&self) -> Result<Snapshot, ()> {
+        // If this is the first time the task is polled, the task will be bound
+        // to the scheduler, in which case the task ref count must be
+        // incremented.
+        let is_not_bound = !self.scheduler.is_bound();
+
+        // Transition the task to the running state.
+        //
+        // A failure to transition here indicates the task has been cancelled
+        // while in the run queue pending execution.
+        let snapshot = match self.header.state.transition_to_running(is_not_bound) {
+            Ok(snapshot) => snapshot,
+            Err(_) => {
+                // The task was shutdown while in the run queue. At this point,
+                // we just hold a ref counted reference. Drop it here.
+                return Err(());
```

Comment:
That said, the comment here is now wrong. It would also be good to use an error type with an appropriate name, analogous to `PollFuture::DropReference`.

Reply:
Added!